gufe/storage/pseudodirectory.py

```diff
@@ -0,0 +1,156 @@
+from typing import Union, Optional
+from pathlib import Path
+from os import PathLike
+from .externalresource import ExternalStorage
+
+import logging
+_logger = logging.getLogger(__name__)
+
+class SharedRoot:
+    """
+    Parameters
+    ----------
+    scratch : os.PathLike
+        the scratch directory shared by all objects on this host
+    external : :class:`.ExternalStorage`
+        external storage resource where objects should eventualy go
+    prefix : str
+        label for this specific unit
+    holding : os.PathLike
+        name of the subdirectory of scratch where shared results are
+        temporarily stored; default is '.holding'. This must be the same for
+        all units within a DAG.
+    delete_holding : bool
+        whether to delete the contents of the $SCRATCH/$HOLDING/$PREFIX
+        directory when this object is deleted
+    read_only : bool
+        write to prevent NEW files from being written within this shared
+        directory. NOTE: This will not prevent overwrite of existing files,
+        in scratch space, but it will prevent changed files from uploading
+        to the external storage.
```

Comment:
This `read_only` is mainly to prevent the following mistake:

```python
shared = root.get_other_shared(other_unit_prefix)
with open(shared / "oldfile.txt") as f:
    ... # good so far!

with open(shared  / "newfile.txt", mode='wb') as f:
    ... # uh-oh -- now we're writing into the wrong unit's space!
```

However, it *won't* stop `open(shared / "oldfile.txt", mode='wb')`. You can still overwrite that data. If that data is in a separate permanent storage, it will be fine -- only overwrite locally. (If you set `shared` up such that the directory it uses is identical to the internal holding cache here, of course, it would overwrite that.)